### PR TITLE
CFE-351

### PR DESCRIPTION
**Pull request by sherine-k**
Main goal:
* All resources created with a nodeobservability CR should be cleaned up upon delete of the CR
Also contains:
* Reduce open rbac permissions of the operator to the namespace scope
* Remove secret from reconcile loop : a secret containing the service account token is always automatically created and mounted on the pod using it

### CODE REVIEW

File: pkg/operator/controller/nodeobservability/controller.go [hunk 1]
```diff
@@ -195,3 +204,85 @@ func (r *NodeObservabilityReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+	if err := r.deleteClusterRole(nodeObs); err != nil {
+		errs = append(errs, fmt.Errorf("failed to delete clusterrole : %w", err))
+	}
+	if err := r.deleteClusterRoleBinding(nodeObs); err != nil {
+		errs = append(errs, fmt.Errorf("failed to delete clusterrolebinding : %w", err))
+	}
+	if err := r.deleteSecurityContextConstraints(nodeObs); err != nil {
+		errs = append(errs, fmt.Errorf("failed to delete SCC : %w", err))
+	}
+	if len(errs) == 0 {
```
Comment (lmzuccarelli) [inline pkg/operator/controller/nodeobservability/controller.go:278]: nit: merge if statement i.e `if len(errs) == 0 && hasFinalizer(nodeObs) {`

**Review by lmzuccarelli — COMMENTED** [review]
I tried deleting the nodeobeservability sample I created and got this error 

`Reconciler error	{"reconciler group": "nodeobservability.olm.openshift.io", "reconciler kind": "NodeObservability", "name": "nodeobservability-sample", "namespace": "node-observability-operator", "error": "failed to update NodeObservability with finalizers:, failed to update finalizers: Operation cannot be fulfilled on nodeobservabilities.nodeobservability.olm.openshift.io \"nodeobservability-sample\": StorageError: invalid object, Code: 4, Key: /kubernetes.io/nodeobservability.olm.openshift.io/nodeobservabilities/node-observability-operator/nodeobservability-sample, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 0563674f-07cc-446f-9eff-b59d2b9c8616, UID in object meta: "}`

Need to look at why the finalizers we not able to update

File: pkg/operator/controller/nodeobservability/rbac.go [hunk 2]
```diff
@@ -84,6 +81,14 @@ func (r *NodeObservabilityReconciler) desiredClusterRole(nodeObs *v1alpha1.NodeO
...
 			Namespace: nodeObs.Namespace,
 			Labels:    labelsForClusterRole(clusterRoleName),
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					Name:       nodeObs.Name,
+					Kind:       nodeObs.Kind,
+					UID:        nodeObs.UID,
+					APIVersion: nodeObs.APIVersion,
+				},
+			},
```
Comment (sherine-k) [inline pkg/operator/controller/nodeobservability/rbac.go:91]: @lmzuccarelli @josefkarasek 
Do you think we should keep this? 
It normally doesnt have any effect to set ownership on cluster scoped resources.
Let me know 
Applicable to SCC and clusterrolebinding as well

File: pkg/operator/controller/nodeobservability/controller.go [hunk 3]
```diff
@@ -91,12 +96,26 @@ func (r *NodeObservabilityReconciler) Reconcile(ctx context.Context, req ctrl.Re
 		r.Log.Error(err, "Failed to get NodeObservability")
 		return ctrl.Result{}, err
 	}
+	if nodeObs.DeletionTimestamp != nil {
+		r.Log.Info("NodeObservability resource is going to be delete. Taking action")
```
Comment (lmzuccarelli) [inline pkg/operator/controller/nodeobservability/controller.go:100]: nit: delete -> deleted

File: pkg/operator/controller/nodeobservability/rbac.go (hunk 2 above)
Comment (lmzuccarelli) [reply pkg/operator/controller/nodeobservability/rbac.go:91]: If it has no effect then delete it - less code is always better :)

File: pkg/operator/controller/nodeobservability/controller.go [hunk 4]
```diff
@@ -195,3 +204,85 @@ func (r *NodeObservabilityReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		Owns(&appsv1.DaemonSet{}).
 		Complete(r)
 }
+func hasFinalizer(nodeObs *operatorv1alpha1.NodeObservability) bool {
```
[the remark this answers is not in the document]
Comment (sherine-k) [reply pkg/operator/controller/nodeobservability/controller.go:209]: done

File: pkg/operator/controller/nodeobservability/rbac.go (hunk 2 above)
Comment (sherine-k) [reply pkg/operator/controller/nodeobservability/rbac.go]: OK done

File: pkg/operator/controller/nodeobservability/controller.go (hunk 3 above)
Comment (sherine-k) [reply pkg/operator/controller/nodeobservability/controller.go]: oops

**Comment by sherine-k** [comment]
> I tried deleting the nodeobeservability sample I created and got this error
> 
> `Reconciler error {"reconciler group": "nodeobservability.olm.openshift.io", "reconciler kind": "NodeObservability", "name": "nodeobservability-sample", "namespace": "node-observability-operator", "error": "failed to update NodeObservability with finalizers:, failed to update finalizers: Operation cannot be fulfilled on nodeobservabilities.nodeobservability.olm.openshift.io \"nodeobservability-sample\": StorageError: invalid object, Code: 4, Key: /kubernetes.io/nodeobservability.olm.openshift.io/nodeobservabilities/node-observability-operator/nodeobservability-sample, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 0563674f-07cc-446f-9eff-b59d2b9c8616, UID in object meta: "}`
> 
> Need to look at why the finalizers we not able to update

I didn't reproduce this one

**Pull request merged**